Pages: Add new download page

harding opened the pull request:
Preview: http://dg0.dtrt.org/en/download/

This PR adds a download page to the site based on the Bitcoin.org DL page historically used for Bitcoin Core.

![new-dl-page](https://user-images.githubusercontent.com/61096/31449994-4ad9d7fa-ae76-11e7-9e62-1df8033a9d10.png)

**Features:**

1. **Easily translatable:** the layout, styling, and links are a template, so all that is needed for localization is translating about two dozen strings.  See `_posts/en/pages/2017-01-01-download.md` for a well-commented example.

2. **Top-level download link:** a link is added to the top-level navigation menu linking directly to the DL page.  We may want to consider removing the Releases page from the top-level nav menu, as it is now linked to from the DL page.

3. **Auto version number:** the version number title and the version numbers used in all file names are automatically set to the highest version number whose release notes file has been added to the site.  This is a tiny bit hokey, as you have to add the release number as separate elements into an array in the release file in order for Jekyll to get wacky version-number-style sorting, e.g. `release: [0, 15, 0, 1]`.  See `_releases/0.15.0.1.md` for an example of release notes extended with metadata.

4. **Releases-only RSS feed:** a new RSS feed is added that only contains release notes.  Adding new release notes to the site automatically updates the feed.

5. **Download button with platform autodetection:** the main download button attempts to detect the user's platfrom from their browser user agent and recommends a download based upon it.  I've tested this using a user-agent switcher, but I don't have any non-Linux or non-x86_64 machines to test with for real.  This feature was developed by @saivann for Bitcoin.org (who also did all the page styling).

6. **Templated filenames:** all filenames are filled in from a template in `_data/binaries.yaml` plus the version number described above.  A test run by Travis CI will fail if the binaries for the current release can't be found on the live BitcoinCore.org server for a particular tested commit (see `contrib/qa/test-binary-availability.sh`).

7. **Optional magnet link:** if a magnet link is set in the release notes metadata for the latest release, this magnet icon will appear next to the torrent link.  If no magnet link is set, the icon will not be shown. See `_releases/0.15.0.1.md` for an example of the metadata.

8. **Wladimir's release key is added to the repo** and linked to here, as it is on the Bitcoin.org DL page.

9. **Global datadir size variable:** the paragraph below includes a variable for the approximate size of the datadir so that we can simply update `_data/stats.yml` to change the size across all localized Download pages without having to re-translate anything.  The size estimate here is a little larger than actual on my machine so we don't have to update this variable too frequently.

Not shown, I've attributed the MIT-licensed content copied from Bitcoin.org to "The Bitcoin.org Project".  This is `_includes/templates/download.html`, `_sass/download.scss`, the files in the directory `assets/images/icons/`, and the files in the directory `assets/images/os/`.  For the image directories, I added a COPYING.md file into the directory.

---

For review, I suggest special attention be paid to getting the strings to say what we want, as once they start getting translated, changing them will require special care.  The current strings are all identical to the ones used on Bitcoin.org, except for the datadir size being variable.

achow101 commented:
ACK a0b2c8d15cc4d71d71356256704ef4a95e64b452

btcdrak commented:
Travis passed but somehow the PR hasnt picked up on it. I also restarted travis with the same result

harding commented:
Rebased to fix Travis (which seems to have worked) and added a commit changing the download page link in the recent Twitter Impersonation post to the new DL page.  For comparison, the pre-rebase history is at https://github.com/harding/website/commits/downloadpage-orig

harding commented on _releases/0.15.0.1.md:
```diff
@@ -9,6 +9,16 @@ lang: en
+## Use array so that sort order is [0.1.1, 0.10.0] rather than opposite
+release: [0, 15, 0, 1]
+
+## Optional magnet link.  To get it, open the torrent in a good BitTorrent client
```
Yes.  The only use of the magnet link is on the /en/dowload page, and that only displays the latest release, so I only copied the magnet link from Bitcoin.org for that release.  If we later add a feature that provides magnet links for historic releases, I can go back and get those from Bitcoin.org or from Wladimir's release emails (or generate them myself from the torrent files).

harding commented on _includes/templates/download.html:
```diff
@@ -0,0 +1,139 @@
+      </div>
+    </div>
+    <p class="downloadmore">
+      <a href="{{ PATH_PREFIX }}/SHA256SUMS.asc" class="dl">{{ page.downloadsig }}</a><br>
```
That's expected, since it points to the /bin directory that is only available on the server.  The Travis CI test included in the PR ensures the URL used in the local build exists on the the live server.

harding commented on _includes/templates/download.html:
```diff
@@ -0,0 +1,139 @@
+    </div>
+    <p class="downloadmore">
+      <a href="{{ PATH_PREFIX }}/SHA256SUMS.asc" class="dl">{{ page.downloadsig }}</a><br>
+      <a href="{{ PATH_PREFIX }}/{{ FILE_PREFIX }}.torrent" class="dl">{{ page.downloadtorrent }}</a>
```
This is the same as the SHA256SUMS.asc file in the previous comment: that file only exists on the server because it is part of the binary upload and not part of the repository.  A test is included in the PR to ensure that the link works.  This applies to all links with the class "dl"---none of them are expected to exist in local builds, all of them are expected to exist on the live server, and all of them are checked by the test included in this PR.

harding commented on contrib/qa/test-binary-availability.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash 
+
+# Local builds and Travis CI builds of the BitcoinCore.org website do
```
Isn't that harder to read?  DNS is case insensitive, so both resolve to the same location, and it's always been my habit to capitalize domain names as proper nouns in paragraph text.  I'm happy to paint this bikeshed any color you want, I'm just a little curious why it has to be hot pink. :-)

jnewbery replied:
no, any colour is fine :)

my general preference is for all lowercase since that seems like the normalized version to me, but if your convention is to uppercase proper nouns that's also fine.

harding commented on contrib/qa/test-binary-availability.sh:
```diff
@@ -0,0 +1,29 @@
+# simply restarting the tests after the binaries have been uploaded
+# should resolve the issue and allow a clean merge.
+#
+# If all URLs work: prints nothing, returns 0.
```
I'm quite willing to do that, but I generally adhere to the [rule of silence](http://www.linfo.org/rule_of_silence.html) and let the exit status do all the talking when there's nothing specific to say.  Let me know if you feel strongly about changing it, and I will.

jnewbery replied:
no, I don't feel strongly. The reason for including output from a successful run is that it means it's obvious when the script has been run successfully and when the script wasn't executed at all for some reason.

harding commented:
Some great feedback from @jnewbery .  Thanks!  I'll edit and push an update later today.

harding commented:
Updated to address @jnewbery's excellent feedback.  Live preview also updated: http://dg0.dtrt.org/en/download/

Below is a screenshot of the updated phrasing on the note about resource requirements.  "180", "6", and "5-10" are all variables that can be changed globally across the site without re-translation.  

![2017-10-12-17_29_35_267123385](https://user-images.githubusercontent.com/61096/31520205-1df14b08-af73-11e7-8fb3-ef81f0f7594d.png)

I can squash the new commits if requested, but I kept them separate for now to avoid anyone needing to re-validate the security elements of this PR.

jnewbery commented:
post-merge ACK. Live site looks great. Thanks @harding!

MarcoFalke commented on _includes/templates/download.html:
```diff
@@ -0,0 +1,139 @@
+      <a href="{{ PATH_PREFIX }}/SHA256SUMS.asc" class="dl">{{ page.downloadsig }}</a><br>
+      <a href="{{ PATH_PREFIX }}/{{ FILE_PREFIX }}.torrent" class="dl">{{ page.downloadtorrent }}</a>
+        {% if magnet %} <a href="{{ magnet | replace: '&', '\&amp;'}}" class="magnetlink" data-proofer-ignore></a>{% endif %}<br>
+      <a href="https://github.com/bitcoin/bitcoin" class="dl">{{ page.source }}</a><br>
```
We might want to link to the generated targz archive of the source code instead. And rename this link to something like "git source".
